File: packages/expo-app-metrics/android/src/main/java/expo/modules/appmetrics/networkrequests/NetworkRequestPersistence.kt
// Copyright 2025-present 650 Industries. All rights reserved.

package expo.modules.appmetrics.networkrequests

import android.util.Log
import expo.modules.appmetrics.storage.MetricsDatabase
import expo.modules.appmetrics.storage.Span
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.json.JSONArray
import org.json.JSONObject

private const val TAG = "ExpoAppMetrics"

/**
 * Routes completed requests from `NetworkRequestMonitor` into the `spans` table, attributed to
 * the main session. This is the first span producer: it converts each request into a generic
 * `Span` following the OTel HTTP semantic conventions, so the export layer (`expo-observe`)
 * ships spans without knowing what produced them. Unlike the in-memory ring buffer, the rows
 * survive process death until they're dispatched, pruned with their session, or displaced past
 * the table's row cap. Mirrors the iOS `NetworkRequestPersistence`.
 *
 * Installed on the monitor once the module created the main session (see `AppMetricsModule`);
 * the monitor calls `persist` directly for every completion it records. Deliberately not a
 * `NetworkRequestObserverDelegate`: persistence is part of the pipeline, not an observer, and
 * the delegate contract (weak registration, `shouldObserveRequest` filtering) doesn't apply.
 */
class NetworkRequestPersistence(
  private val database: MetricsDatabase,
  private val scope: CoroutineScope,
  initialConfiguration: NetworkSpansConfiguration = NetworkSpansConfiguration(),
  /**
   * A plain value, not a provider: the id is constant for a persistence instance (it is
   * constructed after the main session exists), and resolving it eagerly means the monitor's
   * record path (OkHttp dispatcher threads) never calls back into module state that a
   * teardown could have invalidated.
   */
  private val sessionId: String
) {
  /**
   * Capture-time recording policy. Volatile because the monitor calls in from OkHttp dispatcher
   * threads while JS reconfigures from the modules queue. Checked before each insert so a change
   * takes effect for future requests immediately; rows persisted earlier are untouched.
   */
  @Volatile
  private var configuration: NetworkSpansConfiguration = initialConfiguration

  /**
   * Applies a new recording policy to subsequent requests. Called when JS reconfigures
   * `traces.network`; the caller persists the value separately.
   */
  fun setConfiguration(configuration: NetworkSpansConfiguration) {
    this.configuration = configuration
  }

  /**
   * Persists one completed request as a span. The insert runs on `scope` so the monitor's
   * record path (OkHttp dispatcher threads) never blocks on the database; failures are logged
   * and swallowed — persistence must never break the monitor's fan-out.
   */
  fun persist(request: NetworkRequest) {
    if (!configuration.allows(request.url, request.method)) {
      return
    }
    val span = request.toSpan(sessionId) ?: return
    scope.launch {
      try {
        database.spanDao().insert(span)
      } catch (e: Exception) {
        Log.w(TAG, "Failed to persist a network request span", e)
      }
    }
  }

  /**
   * Persists the monitor's buffered startup requests. One coroutine for the whole batch: the
   * install path runs on the module's serial queue during startup, competing with the session
   * INSERT and crash-report processing, so converting up to 200 requests inline (JSON
   * serialization, URL parsing) and queueing one coroutine per row there would be the most
   * expensive way to do it. Conversion and inserts all happen inside the single launch.
   */
  fun persistBuffered(requests: List<NetworkRequest>, onComplete: () -> Unit = {}) {
    if (requests.isEmpty()) {
      return
    }
    scope.launch {
      for (request in requests) {
        if (!configuration.allows(request.url, request.method)) {
          continue
        }
        val span = request.toSpan(sessionId) ?: continue
        try {
          database.spanDao().insert(span)
        } catch (e: Exception) {
          Log.w(TAG, "Failed to persist a network request span", e)
        }
      }
      // Deliberately not in a `finally`: a cancelled batch must not report completion, so the
      // caller can retry the drain on the next install.
      onComplete()
    }
  }
}

/**
 * Builds a span from a completed request snapshot, per the OTel HTTP semantic conventions for a
 * client span: the span is named after the method, a transport failure or a 4xx/5xx response
 * makes it an ERROR, and each redirect hop becomes an `expo.http.redirect` event.
 *
 * The attribute keys mirror the set the ingestion endpoint extracts into dedicated columns
 * (`http.request.method`, `url.full`, `server.address`, ...). `url.full` is redacted here, at
 * the instrumentation, as the conventions require: userinfo credentials and the default
 * sensitive query values never reach disk or the wire (ingestion redacts again as defense in
 * depth). An intentionally canceled request keeps its span but stays UNSET with no
 * `error.type`, per the conventions. `error.type` must stay low-cardinality: the captured
 * exception class for a transport failure, or the bare status code when the response itself
 * was the error; the localized `errorDescription` goes to the status message instead.
 *
 * Returns `null` when the snapshot carries no usable timestamps: without either endpoint of
 * the request window there is nothing to anchor a span to.
 */
internal fun NetworkRequest.toSpan(sessionId: String): Span? {
  val start = timings.fetchStart?.time
  val end = timings.responseEnd?.time
  val durationMs = (timings.totalDuration * 1_000).toLong()
  val resolvedStart = start ?: end?.minus(durationMs) ?: return null
  val resolvedEnd = end ?: start?.plus(durationMs) ?: return null

  val attributes = JSONObject()
  // Case-sensitive per the conventions: an unknown or nonstandard method becomes `_OTHER`
  // (verbatim value preserved in `http.request.method_original`) and names the span `HTTP`,
  // so caller-controlled verbs can't mint unbounded span names.
  val isKnownMethod = method in KNOWN_HTTP_METHODS
  attributes.put("http.request.method", if (isKnownMethod) method else "_OTHER")
  if (!isKnownMethod) {
    attributes.put("http.request.method_original", method)
  }
  val parsedUrl = url.toHttpUrlOrNull()
  attributes.put("url.full", redactedUrlFull(url, parsedUrl))
  parsedUrl?.host?.let { host ->
    attributes.put("server.address", host)
  }
  // `HttpUrl.port` already resolves the scheme default when the URL carries no explicit port.
  parsedUrl?.port?.let { port ->
    attributes.put("server.port", port)
  }
  statusCode?.let { code ->
    attributes.put("http.response.status_code", code)
  }
  semconvProtocolVersion(networkProtocol)?.let { version ->
    attributes.put("network.protocol.version", version)
  }
  requestBytesSent?.let { bytes ->
    attributes.put("http.request.size", bytes)
  }
  responseBytesReceived?.let { bytes ->
    attributes.put("http.response.size", bytes)
  }
  // Not a semconv attribute: a cached response is timestamped and byte-counted like a download,
  // so without this a large cache hit reads as an impossible transfer rate. Omitted when nothing
  // classified the fetch, so its absence never implies a network load.
  fetchType?.let { type ->
    attributes.put("expo.http.resource.fetch_type", type.attributeValue)
  }
  val httpErrorStatus = (statusCode ?: 0) >= 400
  // An intentional cancellation is not a failure: status stays UNSET and `error.type` is
  // never set, per the conventions.
  val resolvedErrorType = if (canceled) {
    null
  } else {
    errorType ?: statusCode?.toString().takeIf { httpErrorStatus }
  }
  resolvedErrorType?.let { value ->
    attributes.put("error.type", value)
  }

  val failed = !canceled && (errorDescription != null || errorType != null || httpErrorStatus)
  // The conventions model redirects as resent spans (`http.request.resend_count`), not events;
  // one span per chain is a deliberate deviation for this pipeline. The event and its URL
  // attributes are `expo.`-namespaced because the naming guidelines advise against putting
  // application-specific names under a semconv namespace like `http.`; the hop's status code
  // reuses the registry attribute it matches. A hop's arrival time becomes the event time;
  // without one the exporter anchors the event to the span start.
  val events = JSONArray()
  for (redirect in redirects) {
    val event = JSONObject()
    event.put("name", "expo.http.redirect")
    // Only a time that actually falls inside the span window can ship: an OTLP event outside its
    // own span has no valid place on the timeline. Dropping the key leaves the exporter to anchor
    // the event to the span start, which is where a clamp would have put it anyway.
    redirect.respondedAtMs?.takeIf { it in resolvedStart..resolvedEnd }?.let { respondedAtMs ->
      event.put("timeMs", respondedAtMs)
    }
    val eventAttributes = JSONObject()
    eventAttributes.put("expo.http.redirect.from", redirect.fromUrl)
    eventAttributes.put("expo.http.redirect.to", redirect.toUrl)
    eventAttributes.put("http.response.status_code", redirect.statusCode)
    event.put("attributes", eventAttributes)
    events.put(event)
  }

  return Span(
    sessionId = sessionId,
    name = if (isKnownMethod) method else "HTTP",
    kind = Span.CLIENT_KIND,
    startTimestampMs = resolvedStart,
    endTimestampMs = resolvedEnd,
    statusCode = if (failed) Span.STATUS_ERROR else null,
    statusMessage = if (failed) errorDescription else null,
    attributes = attributes.toString(),
    events = if (events.length() > 0) events.toString() else null
  )
}

/**
 * The conventions' known method set: RFC 9110's methods plus PATCH and QUERY, matched
 * case-sensitively as the conventions require.
 */
private val KNOWN_HTTP_METHODS = setOf(
  "GET", "HEAD", "POST", "PUT", "DELETE", "CONNECT", "OPTIONS", "TRACE", "PATCH", "QUERY"
)

/**
 * Query parameter names whose values are redacted by default: the conventions' list for
 * `url.full` (`X-Amz-Signature`, `X-Amz-Credential`, `X-Amz-Security-Token`, `sig`,
 * `X-Goog-Signature`) plus the legacy S3 `AWSAccessKeyId` and `Signature` pair. Compared
 * case-insensitively.
 */
private val SENSITIVE_QUERY_PARAMETERS = setOf(
  "awsaccesskeyid",
  "signature",
  "sig",
  "x-amz-signature",
  "x-amz-credential",
  "x-amz-security-token",
  "x-goog-signature"
)

/**
 * `url.full` with userinfo credentials replaced by `REDACTED:REDACTED` and default-sensitive
 * query values replaced by `REDACTED`, per the conventions. Redaction is the instrumentation's
 * job, so secrets never reach the on-device database or the wire.
 */
private fun redactedUrlFull(url: String, parsed: okhttp3.HttpUrl?): String {
  // An unparseable URL can't prove it carries no secrets; drop the query outright. The parsed
  // form is passed in so `toSpan` parses each URL once, since it runs on the OkHttp dispatcher
  // thread for every completed request, and per row in the backfill batch.
  if (parsed == null) {
    return url.substringBefore('?')
  }
  val builder = parsed.newBuilder()
  if (parsed.encodedUsername.isNotEmpty() || parsed.encodedPassword.isNotEmpty()) {
    builder.username("REDACTED")
    builder.password("REDACTED")
  }
  // Rewritten as a string rather than through the builder: every builder route
  // (`removeAllQueryParameters` plus `addQueryParameter`, and `setQueryParameter` alike) appends
  // the parameter to the end of the query, and `url.full` is the key URL-level analysis groups
  // on, so the same endpoint would land under different strings depending on where the signed
  // parameter happened to sit. Splicing the encoded query keeps both the position and the exact
  // encoding of every parameter this function leaves alone.
  parsed.encodedQuery?.let { encodedQuery ->
    val redacted = encodedQuery
      .split('&')
      .joinToString("&") { pair ->
        val separator = pair.indexOf('=')
        // No `=` means a valueless parameter (`?sig`), which carries no secret to redact.
        if (separator < 0) {
          return@joinToString pair
        }
        val name = pair.substring(0, separator)
        if (name.lowercase() in SENSITIVE_QUERY_PARAMETERS) "$name=REDACTED" else pair
      }
    if (redacted != encodedQuery) {
      builder.encodedQuery(redacted)
    }
  }
  return builder.build().toString()
}

/**
 * Bare protocol version per semconv's `network.protocol.version` ("1.1", "2", "3"), mapped from
 * the ALPN-style names OkHttp reports ("http/1.1", "h2", "h3"). Unrecognized values pass through
 * verbatim rather than being dropped.
 */
private fun semconvProtocolVersion(networkProtocol: String?): String? = when {
  networkProtocol == null -> null
  networkProtocol == "h2" -> "2"
  networkProtocol == "h3" -> "3"
  networkProtocol.startsWith("http/") -> networkProtocol.removePrefix("http/")
  else -> networkProtocol
}

File: packages/expo-app-metrics/ios/NetworkRequests/NetworkRequestPersistence.swift
// Copyright 2025-present 650 Industries. All rights reserved.

import Foundation

/// Routes completed requests from `NetworkRequestMonitor` into the `spans` table, attributed to
/// the main session. This is the first span producer: it converts each request into a generic
/// `SpanRow` following the OTel HTTP semantic conventions, so the export layer (`expo-observe`)
/// ships spans without knowing what produced them. Unlike the in-memory ring buffer, the rows
/// survive process death until they're dispatched, pruned with their session, or displaced past
/// the table's row cap.
///
/// Installed on the monitor at launch (see `AppMetricsAppDelegateSubscriber`); the monitor calls
/// `persist` synchronously on `AppMetricsActor` for every completion it records.
@AppMetricsActor
final class NetworkRequestPersistence: Sendable {
  private let writer: SpanWriter
  private let sessionId: @Sendable () -> String

  /// Capture-time recording policy. Checked before each insert so disabling (or filtering)
  /// takes effect for future requests immediately; rows persisted earlier are untouched.
  private var configuration: NetworkSpansConfiguration

  /// `sessionId` is a closure so constructing the persistence doesn't force the session
  /// machinery into existence before the app delegate finished wiring it.
  init(
    writer: SpanWriter,
    configuration: NetworkSpansConfiguration = NetworkSpansConfiguration(),
    sessionId: @escaping @Sendable () -> String
  ) {
    self.writer = writer
    self.configuration = configuration
    self.sessionId = sessionId
  }

  /// Applies a new recording policy to subsequent requests. Called when JS reconfigures
  /// `traces.network`; the caller persists the value separately.
  func setConfiguration(_ configuration: NetworkSpansConfiguration) {
    self.configuration = configuration
  }

  /// Persists one completed request as a span: applies the recording policy, maps the snapshot
  /// onto the generic row shape, and hands it to the shared writer.
  func persist(_ request: NetworkRequest) {
    guard configuration.allows(url: request.url, method: request.method) else {
      return
    }
    guard let row = SpanRow.from(request: request, sessionId: sessionId()) else {
      return
    }
    writer.write(row)
  }
}

extension SpanRow {
  /// Builds a span row from a completed request snapshot, per the OTel HTTP semantic conventions
  /// for a client span: the span is named after the method (`HTTP` for a nonstandard one), a
  /// transport failure or a 4xx/5xx response makes it an ERROR, and each redirect hop becomes an
  /// `expo.http.redirect` event.
  ///
  /// The attribute keys mirror the set the ingestion endpoint extracts into dedicated columns
  /// (`http.request.method`, `url.full`, `server.address`, ...). `url.full` is redacted here, at
  /// the instrumentation, as the conventions require: userinfo credentials and the default
  /// sensitive query values never reach disk or the wire (ingestion redacts again as defense in
  /// depth). An intentionally canceled request keeps its span but stays UNSET with no
  /// `error.type`, per the conventions. RN apps cancel routinely, and counting aborts as errors
  /// would poison every error-rate view. `error.type` must stay low-cardinality: the captured
  /// `domain:code` pair for a transport failure, or the bare status code when the response itself
  /// was the error; the localized `errorDescription` goes to the status message instead.
  ///
  /// Returns `nil` when the snapshot carries no usable timestamps: without either endpoint of
  /// the request window there is nothing to anchor a span to. The capture factory always sets
  /// both, so this only guards directly constructed values.
  static func from(request: NetworkRequest, sessionId: String) -> SpanRow? {
    let start = request.timings.fetchStart
    let end = request.timings.responseEnd
    let duration = request.timings.totalDuration
    guard
      let resolvedStart = start ?? end?.addingTimeInterval(-duration),
      let resolvedEnd = end ?? start?.addingTimeInterval(duration)
    else {
      return nil
    }

    // Case-sensitive per the conventions: an unknown or nonstandard method becomes `_OTHER`
    // (verbatim value preserved in `http.request.method_original`) and names the span `HTTP`,
    // so caller-controlled verbs can't mint unbounded span names.
    let isKnownMethod = knownHttpMethods.contains(request.method)
    var attributes: [String: Any] = [
      "http.request.method": isKnownMethod ? request.method : "_OTHER",
      "url.full": redactedUrlFull(request.url),
    ]
    if !isKnownMethod {
      attributes["http.request.method_original"] = request.method
    }
    if let host = request.url.host {
      attributes["server.address"] = host
    }
    if let port = request.url.port ?? defaultPort(forScheme: request.url.scheme) {
      attributes["server.port"] = port
    }
    if let statusCode = request.statusCode {
      attributes["http.response.status_code"] = statusCode
    }
    if let version = semconvProtocolVersion(request.networkProtocol) {
      attributes["network.protocol.version"] = version
    }
    if let requestBytesSent = request.requestBytesSent {
      attributes["http.request.size"] = requestBytesSent
    }
    if let responseBytesReceived = request.responseBytesReceived {
      attributes["http.response.size"] = responseBytesReceived
    }
    // Not a semconv attribute: a cached response is timestamped and byte-counted like a download,
    // so without this a large cache hit reads as an impossible transfer rate. Omitted when the
    // platform did not classify the fetch, so its absence never implies a network load.
    if let fetchType = request.fetchType {
      attributes["expo.http.resource.fetch_type"] = fetchType.rawValue
    }
    let httpErrorStatus = (request.statusCode ?? 0) >= 400
    // An intentional cancellation is not a failure: status stays UNSET and `error.type` is
    // never set, per the conventions.
    let canceled = request.errorType == canceledErrorType
    if !canceled, let errorType = request.errorType ?? (httpErrorStatus ? request.statusCode.map(String.init) : nil) {
      attributes["error.type"] = errorType
    }

    let failed = !canceled && (request.errorDescription != nil || request.errorType != nil || httpErrorStatus)
    // The conventions model redirects as resent spans (`http.request.resend_count`), not
    // events; one span per chain is a deliberate deviation for this pipeline. The event and its
    // URL attributes are `expo.`-namespaced because the naming guidelines advise against putting
    // application-specific names under a semconv namespace like `http.`; the hop's status code
    // reuses the registry attribute it matches. A hop's arrival time becomes the event time;
    // without one the exporter anchors the event to the span start.
    let events: [[String: Any]] = request.redirects.map { redirect in
      var event: [String: Any] = [
        "name": "expo.http.redirect",
        "attributes": [
          "expo.http.redirect.from": redirect.fromUrl.absoluteString,
          "expo.http.redirect.to": redirect.toUrl.absoluteString,
          "http.response.status_code": redirect.statusCode,
        ],
      ]
      // Only a time that actually falls inside the span window can ship: an OTLP event outside
      // its own span has no valid place on the timeline. Dropping the key leaves the exporter to
      // anchor the event to the span start, which is where a clamp would have put it anyway.
      if let respondedAt = redirect.respondedAt?.unixMilliseconds,
        respondedAt >= resolvedStart.unixMilliseconds,
        respondedAt <= resolvedEnd.unixMilliseconds
      {
        event["timeMs"] = respondedAt
      }
      return event
    }

    return SpanRow(
      sessionId: sessionId,
      name: isKnownMethod ? request.method : "HTTP",
      kind: SpanRow.clientKind,
      startTimestampMs: resolvedStart.unixMilliseconds,
      endTimestampMs: resolvedEnd.unixMilliseconds,
      statusCode: failed ? SpanRow.statusError : nil,
      statusMessage: failed ? request.errorDescription : nil,
      attributes: serializeJSON(attributes),
      events: events.isEmpty ? nil : serializeJSON(events)
    )
  }

  /// The conventions' known method set: RFC 9110's methods plus PATCH and QUERY, matched
  /// case-sensitively as the conventions require.
  private static let knownHttpMethods: Set<String> = [
    "GET", "HEAD", "POST", "PUT", "DELETE", "CONNECT", "OPTIONS", "TRACE", "PATCH", "QUERY",
  ]

  /// `errorType` value NSURLSession reports for an intentional cancellation
  /// (`NSURLErrorCancelled`).
  private static let canceledErrorType = "NSURLErrorDomain:-999"

  /// Query parameter names whose values are redacted by default: the conventions' list for
  /// `url.full` (`X-Amz-Signature`, `X-Amz-Credential`, `X-Amz-Security-Token`, `sig`,
  /// `X-Goog-Signature`) plus the legacy S3 `AWSAccessKeyId` and `Signature` pair. Compared
  /// case-insensitively.
  private static let sensitiveQueryParameters: Set<String> = [
    "awsaccesskeyid", "signature", "sig", "x-amz-signature", "x-amz-credential", "x-amz-security-token",
    "x-goog-signature",
  ]

  /// `url.full` with userinfo credentials replaced by `REDACTED:REDACTED` and default-sensitive
  /// query values replaced by `REDACTED`, per the conventions. Redaction is the
  /// instrumentation's job, so secrets never reach the on-device database or the wire.
  private static func redactedUrlFull(_ url: URL) -> String {
    guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
      // An unparseable URL can't prove it carries no secrets; drop the query outright.
      return url.absoluteString.components(separatedBy: "?").first ?? url.absoluteString
    }
    // Rewrite the components only when something was actually redacted, and go through
    // `percentEncodedQueryItems` rather than `queryItems`: the decoded form is not byte-faithful
    // (`a%2Bb` decodes to `a+b` and re-encodes as a literal `+`, which backends read as a space),
    // and that would corrupt the parameters this function keeps, not just the ones it replaces.
    // The sensitive names contain no characters that percent-encode, so matching the encoded name
    // is equivalent to matching the decoded one.
    var redactedAnything = false
    if components.user != nil || components.password != nil {
      components.user = "REDACTED"
      components.password = "REDACTED"
      redactedAnything = true
    }
    if let queryItems = components.percentEncodedQueryItems, !queryItems.isEmpty {
      var itemsChanged = false
      let redactedItems = queryItems.map { item -> URLQueryItem in
        // A valueless parameter (`?sig`) carries no secret, and giving it one would report a URL
        // the app never sent.
        guard sensitiveQueryParameters.contains(item.name.lowercased()), item.value != nil else {
          return item
        }
        itemsChanged = true
        return URLQueryItem(name: item.name, value: "REDACTED")
      }
      if itemsChanged {
        components.percentEncodedQueryItems = redactedItems
        redactedAnything = true
      }
    }
    guard redactedAnything else {
      return url.absoluteString
    }
    return components.string ?? url.absoluteString
  }

  /// Scheme-default port per semconv's `server.port` (a Required attribute for HTTP client
  /// spans), used when the URL carries no explicit port.
  private static func defaultPort(forScheme scheme: String?) -> Int? {
    switch scheme?.lowercased() {
    case "https", "wss":
      return 443
    case "http", "ws":
      return 80
    default:
      return nil
    }
  }

  /// Bare protocol version per semconv's `network.protocol.version` ("1.1", "2", "3"), mapped
  /// from the ALPN-style names the OS reports ("http/1.1", "h2", "h3"). Unrecognized values pass
  /// through verbatim rather than being dropped.
  private static func semconvProtocolVersion(_ networkProtocol: String?) -> String? {
    switch networkProtocol {
    case nil:
      return nil
    case "h2":
      return "2"
    case "h3":
      return "3"
    case .some(let other):
      return other.hasPrefix("http/") ? String(other.dropFirst("http/".count)) : other
    }
  }

  /// Serializes a JSON-compatible value built above; the inputs are all strings and numbers, so
  /// a failure means a programming error and degrading to `nil` (dropped blob) is safe.
  private static func serializeJSON(_ object: Any) -> String? {
    guard let data = try? JSONSerialization.data(withJSONObject: object) else {
      logger.warn("[AppMetrics] Failed to serialize span attributes")
      return nil
    }
    return String(data: data, encoding: .utf8)
  }
}

extension Date {
  /// Unix-epoch milliseconds, the integer form the spans table and the JS redirect record use.
  var unixMilliseconds: Int64 {
    return Int64((timeIntervalSince1970 * 1_000).rounded())
  }
}
